Write integer of_string from scratch

pjuftring opened the pull request:
While #265 did the job of supporting unsigned integers, it certainly was
no nice implementation. This version adds the same features, i.e.
allowing signed integers (as defined in the lexer) and larger unsigned
values, but is actually written from scratch and doesn't rely on the
builtin of_string anymore. This PR should close #225 if it is accepted.

rossberg-chromium commented:
Hm, this code seems to cause assertion failures on invalid input. That should never happen. Can you change it to raise normal exceptions, preferably the same as the original?

Additional nit: there are plenty of redundant parens in the code.

pjuftring commented:
I removed all redundent parentheses, that I could find, and added a "try" to catch my asserts.

rossberg-chromium commented:
Please do not use assertions like this. Assertions are intended for checking invariants (pre/post-conditions) and thus the internal consistency of code, _not_ for checking the validity of input data. An assertion failure is supposed to signal a program bug.

Also, please do never write catch-all try handlers. That is a guarantee for unintentionally masking unrelated errors. Only match the exceptions you expect to get and intend to handle.

pjuftring commented:
Oh, sorry, I didn't know that. I hope, that I solved it correctly.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
-  let of_string = Rep.of_string
+  (* This implementation allows leading signs and unsigned values *)
+  let of_string x =
+    let len = String.length x in
```
Hint: `let open Rep in` allows to omit the repeated module prefix.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+    let len = String.length x in
+    let power_of_two n = Rep.shift_left Rep.one n in
+    let ten = Rep.of_int 10 in
+    let fail () = failwith "int_of_string" in
```
You could make this more compact by turning it into a `require` function, similar e.g. to that used in `check.ml`.

The string should probably be `"of_string"`.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+        else fail ()
+    in
+    let parse_hex offset sign =
+      let num = ref Rep.zero in
```
The code would probably become shorter (and more idiomatic) if you used recursion instead of a loop.

pjuftring replied:
Don't I have to worry about the stack, if I use recursion here?

rossberg-chromium replied:
Well, it should be tail-recursive, of course. Like so:
```
let rec parse_hex i num =
  if i = len then num
  else if overflow... then fail ()
  else parse_hex (i+1) (logor (shift_left num 4) (of_int (parse_hexdigit x.[i])))
```
Btw, I think you can get rid of the entire sign-specific logic in these functions. For values with negative sign it should suffice to simply check once in the end (at line 257 below) whether the value is negative already.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+    let parse_hex offset sign =
+      let num = ref Rep.zero in
+      for i = offset to len - 1 do
+        if ge_u !num (power_of_two (Rep.bitwidth - 4)) then fail ();
```
I think you can simplify this to `gt_u !num (shr_u minus_zero 4)`

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+        if ge_u !num (power_of_two (Rep.bitwidth - 4)) then fail ();
+        num := Rep.logor (Rep.shift_left !num 4) (Rep.of_int (parse_hexdigit x.[i]));
+      done;
+      if sign && gt_u !num (power_of_two (Rep.bitwidth - 1)) then fail ();
```
Why not simply `sign && lt_s !num zero`?

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+      !num
+    in
+    let parse_dec offset sign =
+      let max_upper, max_lower =
```
These and a couple of other locals are constants, so it's better to move them out of the entire function. The same goes for a couple of helper functions that do not depend on `x`.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+    in
+    let parse_dec offset sign =
+      let max_upper, max_lower =
+        if sign then
```
You can push the `if` into the first operand.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+    let parse_dec offset sign =
+      let max_upper, max_lower =
+        if sign then
+          divrem_u (power_of_two (Rep.bitwidth - 1)) ten
```
Shouldn't this be `divrem_u (shr_u minus_one 1) ten` instead?

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,55 @@ struct
+      !num
+    in
+    let parse_int offset sign =
+      if offset + 3 <= len && String.sub x offset 2 = "0x" then
```
Nit: it's cheaper to just compare `x.[offset]` and x.[offset+1]` than allocating a substring.

rossberg-chromium commented:
Thanks! I left a few comments. Also, it would be good to have a few additional tests, especially for boundary conditions.

pjuftring commented:
I addressed the comments, but there's some stuff I'm not sure about:
* Did I move the right helper functions out of of_int?
* Do I raise the correct exceptions in my helper functions?
* How can I test the bounds of my implementation sufficently, when I can't catch lexer-errors within the test-file?

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +210,49 @@ struct
+      else failwith "unexpected digit"
+
+  let neg_safe x =
+    if lt_s (sub x Rep.one) Rep.minus_one then raise Numerics.IntegerOverflow;
```
Since this is called only once, I'd just inline it. And just use `failwith`, too.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +210,49 @@ struct
+  let of_string x =
+    let open Rep in
+    let len = String.length x in
+    let require b = if not b then failwith "of_int" in
```
Did you mean `"of_string"`?

pjuftring replied:
But if I would move this function out, I would have to call it with the error message, like your implementation did in change.ml. I wanted to avoid this, shall I still change it?

rossberg-chromium replied:
Not sure I follow -- it's not using any local variables, so you should be able to just hoist it as is.

pjuftring replied:
Everytime I call "require" I want it to be the same error message "of_string", but if I move it out, I can't hardcode this message, since this particular error message does only make sense in the scope of my "of_string"-function. I will have to supply the same error string everytime I call the hoisted "require"-function. This seemed a bit redundant to me. This is no big deal, I guess, I simply thought, that if I only have one error message, my solution would be a bit compact.

rossberg-chromium replied:
It's still only a helper for to_string, so no need to generalise, you can leave the string fixed. (Being a helper to a specific function does not imply that it has to be local to that function. In general, ML code tends to use modules as the unit of encapsulation, and it's usually not worth being more fine-grained about it.)

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +210,49 @@ struct
+      if i = len then num
+      else begin
+        require ('0' <= x.[i] && '9' >= x.[i]);
+        let new_digit = of_int (int_of_char x.[i] - 0x30) in
```
Just for symmetry with `parse_hexdigit`, perhaps move this and the previous line to a helper function, too.

rossberg-chromium commented:
Thanks, looks pretty good now, just a few nits left.

As for testing, you can write test files named `xyz.fail.wast`, which the runner expects to fail. If you also want to check the specific error output then you have to create associated expected-output files (though that's perhaps overkill in this case).

hhugo commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +210,49 @@ struct
+      if '0' <= c && '9' >= c then 0x30
+      else if 'a' <= c && 'f' >= c then 0x61 - 10
+      else if 'A' <= c && 'F' >= c then 0x41 - 10
+      else failwith "unexpected digit"
```
fwiw, I find the following easier to read : 
```
let parse_hexdigit = function
  | '0' .. '9' as c ->  Char.code c - Char.code '0'
  | 'a' .. 'f' as c ->  0xa + Char.code c - Char.code 'a' 
  | 'A' .. 'F' as c ->  0xa + Char.code c - Char.code 'A'
  | _                  ->  failwith "unexpected digit"
```

pjuftring replied:
I think, this is a lot prettier than my version @rossberg-chromium shall I copy this?

rossberg-chromium replied:
Yes, looks good.

pjuftring commented:
I added some tests and hoisted `parse_decdigit` and `require`.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -250,7 +249,12 @@ struct
       | '+' -> parse_int 1
-      | '-' -> neg_safe (parse_int 1)
+      | '-' ->
+        begin
```
Nit: you don't need the begin/end here

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -210,22 +210,22 @@ struct
+    | '0' .. '9' as c ->  Char.code c - Char.code '0'
+    | 'a' .. 'f' as c ->  0xa + Char.code c - Char.code 'a'
+    | 'A' .. 'F' as c ->  0xa + Char.code c - Char.code 'A'
+    | _ ->  failwith "unexpected digit"
```
This should be "of_string", too.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -210,22 +210,22 @@ struct
-    if lt_s (sub x Rep.one) Rep.minus_one then raise Numerics.IntegerOverflow;
-    Rep.neg x
+  let parse_decdigit c =
+    if '0' > c || '9' < c then failwith "unexpected digit";
```
Same here.

rossberg-chromium commented:
Thanks! LGTM modulo a last couple of nits
